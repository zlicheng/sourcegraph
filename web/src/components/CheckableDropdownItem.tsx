import CheckIcon from 'mdi-react/CheckIcon'
import React from 'react'
import { DropdownItem, DropdownItemProps } from 'reactstrap'

/**
 * A <DropdownItem /> wrapper that optionally shows a checkmark.
 */
export const CheckableDropdownItem: React.FunctionComponent<
    {
        checked: boolean
    } & DropdownItemProps
> = ({ checked, children, ...props }) => (
    <DropdownItem {...props}>
        <div className="d-flex align-items-start">
            <CheckIcon className={`icon-inline flex-0 small mr-3 ${checked ? '' : 'hidden invisible'}`} />
            <div>{children}</div>
        </div>
    </DropdownItem>
)
