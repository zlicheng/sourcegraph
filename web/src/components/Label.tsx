import React from 'react'
import * as GQL from '../../../shared/src/graphql/schema'
import { contrastingForegroundColor } from '../util/contrastingForegroundColor'

interface Props extends Pick<React.HTMLAttributes<HTMLElement>, 'className' | 'onClick'> {
    label: Pick<GQL.ILabel, 'name' | 'color'>
    colorOnly?: boolean

    tag?: 'span' | 'li' | 'button'
}

/**
 * A label (corresponding to the GraphQL Label type).
 */
export const Label: React.FunctionComponent<Props> = ({
    label: { name, color },
    colorOnly = false,
    tag: Tag = 'span',
    className = '',
    children,
    ...props
}) => (
    <Tag
        {...props}
        className={`badge ${colorOnly ? 'p-3' : ''} ${className}`}
        // tslint:disable-next-line: jsx-ban-props
        style={{ backgroundColor: color, color: contrastingForegroundColor(color) }}
    >
        {!colorOnly && name}
        {children}
    </Tag>
)
