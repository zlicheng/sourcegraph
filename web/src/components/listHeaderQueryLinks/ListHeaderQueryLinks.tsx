import H from 'history'
import React from 'react'
import { Link } from '../../../../shared/src/components/Link'
import { QueryParameterProps } from '../../util/useQueryParameter'
import { threadsQueryMatches, threadsQueryWithValues } from './url'

/** A link in {@link ListHeaderQueryLinks}. */
export interface QueryLink {
    label: string
    queryField: string
    queryValues: string[]
    removeQueryFields?: string[]
    count?: number
    icon?: React.ComponentType<{ className?: string }>
}

export const ListHeaderQueryLink: React.FunctionComponent<
    QueryLink &
        Pick<QueryParameterProps, 'query'> & {
            className?: string
            activeClassName?: string
            inactiveClassName?: string
            location: H.Location
        }
> = ({
    label,
    queryField,
    queryValues,
    removeQueryFields,
    count,
    icon: Icon,
    query,
    className = '',
    activeClassName = 'active',
    inactiveClassName = 'text-muted',
    location,
}) => {
    const queryFieldValues: { [name: string]: null | string[] } = { [queryField]: queryValues }
    if (removeQueryFields) {
        for (const f of removeQueryFields) {
            if (f !== queryField) {
                queryFieldValues[f] = null
            }
        }
    }
    return (
        <Link
            to={urlForThreadsQuery(location, threadsQueryWithValues(query, queryFieldValues))}
            className={`${className} ${
                queryValues.every(queryValue => threadsQueryMatches(query, { [queryField]: queryValue }))
                    ? activeClassName
                    : inactiveClassName
            }`}
        >
            {Icon && <Icon className="icon-inline mr-1" />}
            {count} {label}
        </Link>
    )
}

interface Props extends Pick<QueryParameterProps, 'query'> {
    /** The links to display. */
    links: QueryLink[]

    className?: string
    itemClassName?: string
    itemActiveClassName?: string
    itemInactiveClassName?: string

    location: H.Location
}

/**
 * A button group with links that interact with the query that determines the contents of a list,
 * such as showing "4 open" and "3 closed".
 */
export const ListHeaderQueryLinksButtonGroup: React.FunctionComponent<Props> = ({
    query,
    links,
    location,
    className = '',
    itemClassName = '',
    itemActiveClassName = '',
    itemInactiveClassName = '',
}) => (
    <div className={`btn-group ${className}`}>
        {links.map((linkProps, index) => (
            <ListHeaderQueryLink
                key={index}
                {...linkProps}
                query={query}
                className={`btn ${itemClassName}`}
                activeClassName={itemActiveClassName}
                inactiveClassName={itemInactiveClassName}
                location={location}
            />
        ))}
    </div>
)

/**
 * A nav with links that interact with the query that determines the contents of a list, such as
 * showing "4 open" and "3 closed".
 */
export const ListHeaderQueryLinksNav: React.FunctionComponent<Props> = ({
    query,
    links,
    location,
    className = '',
    itemClassName = '',
    itemActiveClassName = '',
    itemInactiveClassName = '',
}) => (
    <ul className={`nav ${className}`}>
        {links.map((linkProps, index) => (
            <li key={index} className="nav-item">
                <ListHeaderQueryLink
                    {...linkProps}
                    query={query}
                    className={`nav-link px-2 ${itemClassName}`}
                    activeClassName={itemActiveClassName}
                    inactiveClassName={itemInactiveClassName}
                    location={location}
                />
            </li>
        ))}
    </ul>
)

function urlForThreadsQuery(location: Pick<H.Location, 'search'>, threadsQuery: string): H.LocationDescriptor {
    const parameters = new URLSearchParams(location.search)
    parameters.set('q', threadsQuery)
    return { ...location, search: `${parameters}` }
}
