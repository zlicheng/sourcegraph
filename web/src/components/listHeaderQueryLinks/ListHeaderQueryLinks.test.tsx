import React from 'react'
import renderer from 'react-test-renderer'
import { setLinkComponent } from '../../../../shared/src/components/Link'
import { ListHeaderQueryLinksNav } from './ListHeaderQueryLinks'

// tslint:disable: jsx-no-lambda
describe('ListHeaderQueryLinks', () => {
    setLinkComponent((props: any) => <a {...props} />)
    afterAll(() => setLinkComponent(null as any)) // reset global env for other tests

    test('simple', () =>
        expect(
            renderer
                .create(
                    <ListHeaderQueryLinksNav
                        query="is:b"
                        links={[
                            { label: 'a', queryField: 'is', queryValues: ['a'], count: 1 },
                            { label: 'b', queryField: 'is', queryValues: ['b'], count: 2 },
                        ]}
                        location={{ pathname: '/', hash: '', state: '', search: 'a=b' }}
                    />
                )
                .toJSON()
        ).toMatchSnapshot())
})
