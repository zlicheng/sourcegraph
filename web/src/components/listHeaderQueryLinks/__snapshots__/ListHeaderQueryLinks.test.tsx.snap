// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListHeaderQueryLinks simple 1`] = `
<ul
  className="nav "
>
  <li
    className="nav-item"
  >
    <a
      className="nav-link  "
      to={
        Object {
          "hash": "",
          "pathname": "/",
          "search": "a=b&q=is%3Aa",
          "state": "",
        }
      }
    >
      1
       
      a
    </a>
  </li>
  <li
    className="nav-item"
  >
    <a
      className="nav-link  "
      to={
        Object {
          "hash": "",
          "pathname": "/",
          "search": "a=b&q=is%3Ab",
          "state": "",
        }
      }
    >
      2
       
      b
    </a>
  </li>
</ul>
`;
